### PR TITLE
tuple_vector - a new STL-compatible container that abstracts a SoA-style data layout

### PR DESCRIPTION
**Pull request by CypherSignal**
Introduction to tuple_vector

`tuple_vector` is a data container that is designed to abstract and simplify the handling of a "structure of arrays" layout of data in memory. In particular, it mimics the interface of `vector`, including functionality to do inserts, erases, push_backs, and random-access. It also provides a `RandomAccessIterator` and corresponding functionality, making it compatible with most STL (and STL-esque) algorithms such as ranged-for loops, `find_if`, `remove_if`, or `sort`.

When used or applied properly, this container can improve performance of some algorithms through cache-coherent data accesses or allowing for sensible SIMD programming, while keeping the structure of a single container, to permit a developer to continue to use existing algorithms in STL and the like.

More information can be found in doc/bonus/tuple_vector_readme.md

### CODE REVIEW

**Comment by CypherSignal** [comment]
Will investigate+fix compilation problems on clang/gcc. I had (obviously) only been testing on MSVC 2017.

**Comment by CypherSignal** [comment]
Further investigation on the MSVC 15 compilation issue is blocked for now, as I attempt to get my system set up for appropriate testing.

**Comment by CypherSignal** [comment]
Travis-ci had a hiccup. Everything else built A-OK, it just had a test case with a spurious failure.

File: include/EASTL/bonus/tuple_vector.h [hunk 1]
```diff
@@ -0,0 +1,1396 @@
...
+	typedef eastl_size_t size_type;
+
+	// functions that get piped through swallow need to return some kind of value, hence why these are not void
+	int DoUninitializedMoveAndDestruct(const size_type begin, const size_type end, T* pDest)
+	{
+		T* pBegin = mpData + begin;
+		T* pEnd = mpData + end;
+		eastl::uninitialized_move_ptr_if_noexcept(pBegin, pEnd, pDest);
+		eastl::destruct(pBegin, pEnd);
+		return 0;
```
[the remark this answers is not in the document]
Comment (rparolin) [reply include/EASTL/bonus/tuple_vector.h]: you can use the comma operator to make the function call in swallow and then place an integer constant in the expansion expression inorder to make shallow work.

[the remark this answers is not in the document]
Comment (CypherSignal) [reply include/EASTL/bonus/tuple_vector.h]: Ah, I hadn't thought about that solution to the problem. To be honest, this setup was directly inspired by a similar "swallow" arrangement in tuple.h - it's just that I call swallow about 10x more here. That'll be a good cleanup, and to get things more in line for when cpp14 support is dropped.

**Comment by CypherSignal** [comment]
Commits above should resolve some of the issues raised. init_list support for ctor, operator=, assign, and insert (same funcs as used for vector) I'll take care of a bit later.

**Comment by rparolin** [comment]
Another thing I noticed if you can use compressed_pair on the data pointer and the allocator object to get benefit from the size reduction when our allocators are stateless (via the empty base class optimization)... but not a gating thing.   Can be a follow-up PR.

Looks great.  Thanks for the awesome work of providing docs, benchmarks, unit tests, and a fixed_* variant implementation.  Definitely a model PR if I've ever seen one.

File: include/EASTL/bonus/tuple_vector.h [hunk 2]
```diff
@@ -0,0 +1,1355 @@
...
+	T* mpData = nullptr;
+};
+
+// swallow allows for parameter pack expansion of arguments as means of expanding operations performed
+// if a void function is used for operation expansion, it should be wrapped in (..., 0) so that the compiler
+// thinks it has a parameter to pass into the function
+template <typename... Ts>
+void swallow(Ts&&...) { }
+
+inline bool variadicAnd(bool cond) { return cond; }
```
Comment (rparolin) [inline include/EASTL/bonus/tuple_vector.h]: I think you can replace with this:
https://en.cppreference.com/w/cpp/types/conjunction

I implemented a version for EASTL.

File: include/EASTL/bonus/tuple_vector.h [hunk 3]
```diff
@@ -0,0 +1,1355 @@
...
+	}
+
+	template <typename Iterator, typename = typename enable_if<is_iterator_wrapper<Iterator>::value, bool>::type>
+	int validate_iterator(Iterator iter) const EA_NOEXCEPT { return validate_iterator(unwrap_iterator(iter)); }
+
+	template <typename Iterator, typename = typename enable_if<is_iterator_wrapper<Iterator>::value, bool>::type>
+	static bool validate_iterator_pair(Iterator first, Iterator last) EA_NOEXCEPT { return validate_iterator_pair(unwrap_iterator(first), unwrap_iterator(last)); }
+
+protected:
+	allocator_type mAllocator;
```
Comment (rparolin) [inline include/EASTL/bonus/tuple_vector.h]: potentially usage of compressed_pair to remove the memory overhead of an allocator of zero bytes.

**Comment by CypherSignal** [comment]
Thanks for the merge! As mentioned before I’ll submit the init-list as an extra change later, ditto the small change to use compressed-pair.

I’m not sure conjunction would apply here, though - that seems more used for typetraits and TMP, and not more dynamic usages (eg instead of fold expression with &&, since I wanted to keep this on cpp14).

**Pull request merged**